spec/compiler/semantic/macro_spec.cr

```diff
@@ -245,20 +245,146 @@ describe "Semantic: macro" do
       CR
   end
 
-  it "expands macro with block and argument to yield" do
-    assert_type(<<-CR) { int32 }
-      macro foo
-        {{yield 1}}
-      end
+  describe "yield" do
+    it "expands macro expression in def body in block" do
+      assert_type(<<-CR) { int32 }
+        macro foo
+          {{yield 1}}
+        end
 
-      foo do |value|
-        def bar
-          {{value}}
+        foo do |value|
+          def bar
+            {{value}}
+          end
         end
-      end
 
-      bar
-      CR
+        bar
+        CR
+    end
+
+    it "expands macro expression with call" do
+      assert_type <<-CR { string }
+        macro foo
+          {{ yield "foo" }}
+        end
+
+        foo {|x| {{ x.upcase }} }
+        CR
+    end
+
+    it "returns ASTNode" do
+      assert_type <<-CR { string }
+        macro foo
+          {{ (yield "foo").downcase }}
+        end
+
+        foo {|x| {{ x.upcase }} }
+        CR
+    end
+
+    it "return value gets assigned to macro variable" do
+      assert_type <<-CR { uint8 }
+        macro foo
+          {% x = yield "1" %}
+          {{ x }}
+        end
+
+        # FIXME: The macro begin...end is a general workaround, not related to yield
+        foo {|x| {% begin %}{{ x.id }}_u8{% end %} }
```

Comment:
This is what I'm talking about. You don't need to do this in regular macro code, but you always need to do it in block macros. I think this is inconsistent.

Reply:
If anything, I would expect the block body to be macro code already. That is, you do x.upcase inside the block's body, and that is what's returned to the caller.

Reply:
I know this would be inconsistent with something else. But that's my point: I don't think we should do anything about this. We should simply disallow yield arguments in macros.

Reply:
This begin...raise is necessary in regular macro code, unless you're already in an explicit macro context (like the body of a macro def).

```cr
{{ 1 }}_u8 # Error: unexpected token: "_u8"
```

I suppose we could consider the block body of a macro yield as an explicit macro context as well. Not sure about other implications. But it's probably fine.

> If anything, I would expect the block body to be macro code already. That is, you do x.upcase inside the block's body, and that is what's returned to the caller.

That would be very limiting. The idea is that `{{ yield }}` can expand to *any* code, including macro and non-macro code.